### PR TITLE
CMP-2614: Implement update timestamps on ComplianceCheckResults

### PR DESCRIPTION
**Pull request by Vincent056**
Adding the lastscanned timestamp annotation in each `complianceCheckResult`, added annotation `compliance.openshift.io/last-scanned-timestamp` with a value equal to `scan.status.StartTimestamp`

example of CCR
```yaml
kind: ComplianceCheckResult
metadata:
  annotations:
    compliance.openshift.io/last-scanned-timestamp: "2024-08-15T04:26:11Z"
    compliance.openshift.io/rule: scc-limit-root-containers
  creationTimestamp: "2024-08-15T04:26:43Z"
  generation: 1
  labels:
    compliance.openshift.io/check-severity: medium
    compliance.openshift.io/check-status: MANUAL
    compliance.openshift.io/profile-guid: a230315d-3e4a-5b58-b00f-f96f1553e036
    compliance.openshift.io/scan-name: ocp4-cis
    compliance.openshift.io/suite: ocp4-cis-ssb
  name: ocp4-cis-scc-limit-root-containers
  namespace: openshift-compliance
```
```yaml
[vincent@node compliance-operator]$ oc get scan ocp4-cis -o yaml
apiVersion: compliance.openshift.io/v1alpha1
kind: ComplianceScan
metadata:
  creationTimestamp: "2024-08-15T04:26:10Z"
  finalizers:
  - scan.finalizers.compliance.openshift.io
  generation: 1
  labels:
    compliance.openshift.io/profile-guid: a230315d-3e4a-5b58-b00f-f96f1553e036
    compliance.openshift.io/suite: ocp4-cis-ssb
  name: ocp4-cis
  namespace: openshift-compliance
  ownerReferences:
  - apiVersion: compliance.openshift.io/v1alpha1
    blockOwnerDeletion: true
    controller: true
    kind: ComplianceSuite
    name: ocp4-cis-ssb
    uid: b80577a9-de77-4b7f-af6c-cfb7eee7e36f
  resourceVersion: "5290391"
  uid: 7abb45d7-1a61-491a-bbc1-306406d75e04
spec:
  content: ssg-ocp4-ds.xml
  contentImage: ghcr.io/complianceascode/k8scontent:latest
  maxRetryOnTimeout: 3
  profile: xccdf_org.ssgproject.content_profile_cis
  rawResultStorage:
    nodeSelector:
      node-role.kubernetes.io/master: ""
    pvAccessModes:
    - ReadWriteOnce
    rotation: 3
    size: 1Gi
    tolerations:
    - effect: NoSchedule
      key: node-role.kubernetes.io/master
      operator: Exists
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    - effect: NoSchedule
      key: node.kubernetes.io/memory-pressure
      operator: Exists
  scanTolerations:
  - operator: Exists
  scanType: Platform
  showNotApplicable: false
  strictNodeScan: true
  timeout: 30m
status:
  conditions:
  - lastTransitionTime: "2024-08-15T04:26:14Z"
    message: Compliance scan run is running the scans
    reason: Running
    status: "True"
    type: Processing
  - lastTransitionTime: "2024-08-15T04:26:11Z"
    message: Compliance scan doesn't have results yet
    reason: Processing
    status: "False"
    type: Ready
  phase: AGGREGATING
  remainingRetries: 3
  result: NOT-AVAILABLE
  resultsStorage:
    name: ocp4-cis
    namespace: openshift-compliance
  startTimestamp: "2024-08-15T04:26:11Z"
```

### CODE REVIEW

**Comment by xiaojiey** [comment]
```
 vet: cmd/manager/aggregator.go:346:6: no new variables on left side of :=
make: *** [Makefile:308: vet] Error 1
```

**Comment by xiaojiey** [comment]
verification pass with CATALOG_IMG=ghcr.io/complianceascode/compliance-operator-catalog:591-2ea830149c4ab084010d8c5544a0ac54dca8542c:
```
% oc get scan
NAME                    PHASE   RESULT
ocp4-stig               DONE    NON-COMPLIANT
ocp4-stig-node-master   DONE    NON-COMPLIANT
ocp4-stig-node-worker   DONE    NON-COMPLIANT
rhcos4-stig-master      DONE    NON-COMPLIANT
rhcos4-stig-worker      DONE    NON-COMPLIANT
% oc get scan -o=jsonpath="{.items[*].status.startTimestamp}"
2024-08-15T08:03:25Z 2024-08-15T08:03:24Z 2024-08-15T08:03:25Z 2024-08-15T08:03:25Z 2024-08-15T08:03:23Z%                                                                               
% oc get ccr ocp4-stig-accounts-restrict-service-account-tokens -o=jsonpath={.metadata.annotations}| jq -r        
{
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
  "compliance.openshift.io/rule": "accounts-restrict-service-account-tokens"
}
% oc get ccr -o=jsonpath="{.items[*].metadata.annotations}" | jq -r | grep last-scanned-timestamp | head -n 5
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:03:25Z",
% oc get ccr -o=jsonpath="{.items[*].metadata.annotations}" | jq -r | grep last-scanned-timestamp | wc -l     
     512
% oc get ccr --no-headers | wc -l
 % oc annotate compliancescan/ocp4-stig compliance.openshift.io/rescan=
compliancescan.compliance.openshift.io/ocp4-stig annotated
 % oc get scan -w
NAME                    PHASE     RESULT
ocp4-stig               RUNNING   NOT-AVAILABLE
ocp4-stig-node-master   DONE      NON-COMPLIANT
ocp4-stig-node-worker   DONE      NON-COMPLIANT
rhcos4-stig-master      DONE      NON-COMPLIANT
rhcos4-stig-worker      DONE      NON-COMPLIANT
ocp4-stig               AGGREGATING   NOT-AVAILABLE
ocp4-stig               DONE          NON-COMPLIANT
^C%                                                                                                                                                                                      % oc get scan -o=jsonpath="{.items[*].status.startTimestamp}"                                             
2024-08-15T08:22:39Z 2024-08-15T08:03:24Z 2024-08-15T08:03:25Z 2024-08-15T08:03:25Z 2024-08-15T08:03:23Z%                                                                   
% oc get ccr ocp4-stig-accounts-restrict-service-account-tokens -o=jsonpath={.metadata.annotations}| jq -r        
{
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T08:22:39Z",
  "compliance.openshift.io/rule": "accounts-restrict-service-account-tokens"
}
```

File: cmd/manager/aggregator.go [hunk 1]
```diff
@@ -336,6 +336,31 @@ func createOrUpdateOneResult(crClient aggregatorCrClient, owner metav1.Object, l
...
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetriesForTimestamp))
+	if err != nil {
+		cmdLog.Error(err, "Failed to get scan", "ComplianceScan.Name", owner.GetName())
+		return annotations
+	}
+	if annotations == nil {
+		annotations = make(map[string]string)
+	}
+
+	annotations[compv1alpha1.LastScannedTimestampAnnotation] = scan.Status.StartTimestamp.Format(time.RFC3339)
```
Comment (rhmdnd) [inline cmd/manager/aggregator.go:360]: Do we want to allow the last scanned time to be *before* the creation time?

```
$ oc get ccr ocp4-cis-node-worker-file-groupowner-multus-conf  -o json | jq .metadata.annotations
{
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-15T22:17:08Z",
  "compliance.openshift.io/rule": "file-groupowner-multus-conf"
}
$ oc get ccr ocp4-cis-node-worker-file-groupowner-multus-conf  -o json | jq .metadata.creationTimestamp
"2024-08-15T22:17:40Z"
```

File: cmd/manager/aggregator.go [hunk 2]
```diff
@@ -301,13 +301,13 @@ type compResultIface interface {
 
 func createOrUpdateOneResult(crClient aggregatorCrClient, owner metav1.Object, labels map[string]string, annotations map[string]string, exists bool, res compResultIface) error {
 	kind := res.GetObjectKind()
-
```
Comment (rhmdnd) [inline cmd/manager/aggregator.go:304]: nit: unrelated whitespace change?

File: cmd/manager/aggregator.go [hunk 3]
```diff
@@ -301,13 +301,13 @@ type compResultIface interface {
...
 	if err := controllerutil.SetControllerReference(owner, res, crClient.getScheme()); err != nil {
 		cmdLog.Error(err, "Failed to set ownership", "kind", kind.GroupVersionKind().Kind)
 		return err
 	}
 
 	res.SetLabels(labels)
+	annotations = setScanTimestamp(crClient, owner, annotations)
```
Comment (rhmdnd) [inline cmd/manager/aggregator.go:310]: Seems like we have some inconsistency here in using receivers versus functions. What about:

```go
res.setTimestampAnnotations(...)
```

Or reusing the annotation function below?

**Review by rhmdnd — COMMENTED** [review]
A few initial comments on the approach. Looking like it's on the right track, thanks!

File: cmd/manager/aggregator.go (hunk 1 above)
Comment (Vincent056) [reply cmd/manager/aggregator.go:360]: that's a good question, but we need to have a way to associate scan with CCR, and we will not have endTimeStamp at time of updating/creating CCR

File: cmd/manager/aggregator.go (hunk 2 above)
Comment (Vincent056) [reply cmd/manager/aggregator.go:304]: this might be from the autoformatting from my IDE

File: cmd/manager/aggregator.go (hunk 1 above)
Comment (rhmdnd) [reply cmd/manager/aggregator.go:361]: I suppose if we waited until we had an endtime stamp from the ComplianceScan object we'd need to reiterate the entire list of CCRs again to set the `last-scanned-timestamp`.

File: tests/e2e/framework/common.go [hunk 4]
```diff
@@ -1151,6 +1151,43 @@ func (f *Framework) AssertScanIsCompliant(name, namespace string) error {
...
+			return fmt.Errorf("Check %s has no annotations", check.Name)
+		}
+		if annotations[compv1alpha1.LastScannedTimestampAnnotation] == "" {
+			return fmt.Errorf("Check %s has no timestamp annotation", check.Name)
+		}
+		timestamp, err := time.Parse(time.RFC3339, annotations[compv1alpha1.LastScannedTimestampAnnotation])
+		if err != nil {
+			return fmt.Errorf("Error parsing timestamp for check %s: %v", check.Name, err)
+		}
+		if timestamp.Before(cs.Status.StartTimestamp.Time) || timestamp.After(cs.Status.EndTimestamp.Time) {
```
Comment (rhmdnd) [inline tests/e2e/framework/common.go:1184]: If I'm understanding `setTimestampAnnotations` correctly we won't need to apply an OR here. Aren't we always setting the check result time stamp to be the same exact time as the start time for the scan?

That should allow us to be a little tighter in our assertion by checking for the exact time, instead of a range.

File: tests/e2e/parallel/main_test.go [hunk 5]
```diff
@@ -651,6 +651,46 @@ func TestSingleScanSucceeds(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// assertComplianceCheckResultTimestamps checks that the timestamps are set
+	// and that they are set to the same value of startTimestamp of the scan
+	err = f.AssertComplianceCheckResultTimestamps(scanName, f.OperatorNamespace)
+	if err != nil {
+		t.Fatal(err)
+	}
```
Comment (rhmdnd) [inline tests/e2e/parallel/main_test.go:690]: Could we include another case in this test that invokes a rerun, then asserts the scan time is updated?

File: cmd/manager/aggregator.go [hunk 6]
```diff
@@ -336,6 +337,31 @@ func createOrUpdateOneResult(crClient aggregatorCrClient, owner metav1.Object, l
...
+			Name:      owner.GetName(),
+		}, scan)
+		if err != nil {
+			return err
+		}
+		return nil
+	}, backoff.WithMaxRetries(backoff.NewExponentialBackOff(), maxRetriesForTimestamp))
+	if err != nil {
+		cmdLog.Error(err, "Failed to get scan", "ComplianceScan.Name", owner.GetName())
+		return annotations
```
Comment (rhmdnd) [inline cmd/manager/aggregator.go:355]: Won't this result in a nil pointer exception if we don't have any annotations passed in? Should lines 357 - 359 be moved to the top of the function so we guarantee we always return a map of strings, even if we can't get the scan?

File: cmd/manager/aggregator.go (hunk 1 above)
Comment (Vincent056) [reply cmd/manager/aggregator.go:361]: but how would are we able to track down if a scan is linked to a scan that way, don't we need some way to be able to have a unique idenfier for CCR and Scan object, in this case a start timestamp

Comment (Vincent056) [reply cmd/manager/aggregator.go:361]: But I do see the point that it might not make sense to user that the last timestamp is before the creation time

**Comment by xiaojiey** [comment]
@Vincent056 I can still see the last-scanned-timestamp is the same with scan's startTimestamp with index image ghcr.io/complianceascode/compliance-operator-catalog:591-cd850a508d36f939c8eb8cf84c1043577b94cc5f. Is it expected? Thanks.
```
% oc get scan
NAME                    PHASE   RESULT
ocp4-stig               DONE    NON-COMPLIANT
ocp4-stig-node-master   DONE    NON-COMPLIANT
ocp4-stig-node-worker   DONE    NON-COMPLIANT
rhcos4-stig-master      DONE    NON-COMPLIANT
rhcos4-stig-worker      DONE    NON-COMPLIANT
 % oc get scan -o=jsonpath="{.items[*].status.startTimestamp}"                                                
2024-08-26T10:11:23Z 2024-08-26T10:11:27Z 2024-08-26T10:11:27Z 2024-08-26T10:11:23Z 2024-08-26T10:11:23Z%                                                                                                           % oc get ccr -o=jsonpath="{.items[*].metadata.annotations}" | jq -r | grep last-scanned-timestamp | head -n 5
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
% oc get ccr ocp4-stig-accounts-restrict-service-account-tokens -o=jsonpath={.metadata.annotations}| jq -r   
{
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-26T10:11:23Z",
  "compliance.openshift.io/rule": "accounts-restrict-service-account-tokens"
}
```

**Comment by xiaojiey** [comment]
/open
sorry, close it by accident.

File: cmd/manager/aggregator.go (hunk 6 above)
Comment (rhmdnd) [reply cmd/manager/aggregator.go:355]: This should be addressed in https://github.com/ComplianceAsCode/compliance-operator/pull/591/files#diff-37d868832c42510177fe527ffee41dc25b205168859c7a1467e2b9f592fb4ed3L531

File: cmd/manager/aggregator.go [hunk 7]
```diff
@@ -336,6 +337,31 @@ func createOrUpdateOneResult(crClient aggregatorCrClient, owner metav1.Object, l
 	return nil
 }
 
+func setTimestampAnnotations(crClient aggregatorCrClient, owner metav1.Object, annotations map[string]string) map[string]string {
+	// Get the scan timestamp
+	scan := &compv1alpha1.ComplianceScan{}
```
Comment (rhmdnd) [inline cmd/manager/aggregator.go:342]: From a network chatter perspective, we should only need to get this value once before we start creating or update each `ComplianceCheckResult`, right?

If we bury this in the `createOrUpdateOneResult` implementation I think we'll be fetching the scan start time for each result.

Comment (Vincent056) [reply cmd/manager/aggregator.go:342]: that's a good idea, fixing it

**Comment by xiaojiey** [comment]
Verification pass with 4.17.0-0.nightly-2024-08-28-171119 + ghcr.io/complianceascode/compliance-operator-catalog:591-baf525c3eb6cc0f0d80e31a58dadddfb425a8231 index image.
BTW, the ssb will will take much less time than before(previous commint in this PR), especially at AGGREGATING phase.
```
% cat ~/func/ssb_cis.yaml 
apiVersion: compliance.openshift.io/v1alpha1
kind: ScanSettingBinding
metadata:
  name: cis-compliance
  namespace: openshift-compliance
profiles:
  - name: ocp4-cis
    kind: Profile
    apiGroup: compliance.openshift.io/v1alpha1
  - name: ocp4-cis-node
    kind: Profile
    apiGroup: compliance.openshift.io/v1alpha1
settingsRef:
  name: default
  kind: ScanSetting
  apiGroup: compliance.openshift.io/v1alpha1
% oc apply -f ~/func/ssb_cis.yaml
scansettingbinding.compliance.openshift.io/cis-compliance created
% oc get suite -w
NAME             PHASE         RESULT
cis-compliance   AGGREGATING   NOT-AVAILABLE
cis-compliance   DONE          NON-COMPLIANT
cis-compliance   DONE          NON-COMPLIANT
^C%                                                                                                                                                                                                            % date
Thu Aug 29 20:43:59 CST 2024
% oc get ccr ocp4-cis-accounts-restrict-service-account-tokens  -o=jsonpath={.metadata.annotations}| jq -r   
{
  "compliance.openshift.io/last-scanned-timestamp": "2024-08-29T12:42:57Z",
  "compliance.openshift.io/rule": "accounts-restrict-service-account-tokens"
}
```

File: cmd/manager/aggregator.go (hunk 1 above)
Comment (rhmdnd) [reply cmd/manager/aggregator.go:365]: To summarize - the CCR scan timestamps will match the ComplianceScan timestamp. The creation time of the CCR will potentially be later than the scan timestamp. This should only be the case on the first scan, since rescans will cause the CCR to have an updated scan time.

We can highlight that detail in CRD documentation.

File: cmd/manager/aggregator.go (hunk 7 above)
Comment (rhmdnd) [reply cmd/manager/aggregator.go:346]: Aren't we still fetching the ComplianceScan with each ComplianceCheckResult?

File: cmd/manager/common.go [hunk 8]
```diff
@@ -25,7 +26,8 @@ import (
 )
 
 const (
-	maxRetries = 15
+	maxRetries             = 15
+	maxRetriesForTimestamp = 3
```
Comment (rhmdnd) [inline cmd/manager/common.go:30]: nit: We should be able to remove this now since we're not using it anymore.

**Review by rhmdnd — APPROVED** [review]
/lgtm

**Comment by rhmdnd** [comment]
Adding px-approved since we've walked through this with @mkumku.
